File: crates/cxx-qt-lib/src/gui/qpen.cpp
// clang-format off
// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#include "cxx-qt-lib/qpen.h"

#include <cxx-qt-lib/assertion_utils.h>

// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qpen.h?h=v5.15.6-lts-lgpl#n124
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qpen.h?h=v6.2.4#n94
assert_alignment_and_size(QPen, { ::std::size_t a0; });

static_assert(!::std::is_trivially_copy_assignable<QPen>::value);
static_assert(!::std::is_trivially_copy_constructible<QPen>::value);

static_assert(!::std::is_trivially_destructible<QPen>::value);

static_assert(QTypeInfo<QPen>::isRelocatable);

File: tests/qt_types_standalone/rust/src/qpen.rs
// SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QPen;

#[cxx::bridge]
mod qpen_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qpen.h");

        type QPen = cxx_qt_lib::QPen;
    }

    extern "Rust" {
        fn construct_qpen() -> QPen;
        fn clone_qpen(p: &QPen) -> QPen;
    }
}

fn construct_qpen() -> QPen {
    QPen::default()
}

fn clone_qpen(p: &QPen) -> QPen {
    p.clone()
}
